src/protocols/echo/Echo.h

```diff
@@ -0,0 +1,136 @@
+
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <protocols/CHIPProtocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+
+namespace chip {
+namespace Protocols {
+
+enum
+{
+    kEchoMessageType_EchoRequest  = 1,
+    kEchoMessageType_EchoResponse = 2
+};
+
+class DLL_EXPORT EchoClient : public ExchangeContextDelegate
+{
+public:
+    EchoClient(void);
+
+    /**
+     *  Initialize the EchoClient object. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    exchangeMgr    A pointer to the ExchangeManager object.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     */
+    CHIP_ERROR Init(ExchangeManager * exchangeMgr);
+
+    /**
+     *  Shutdown the EchoClient. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     *  @return #CHIP_NO_ERROR unconditionally.
+     */
+    CHIP_ERROR Shutdown(void);
+
+    /**
+     * Send an echo request to a CHIP node.
+     *
+     * @param nodeId        The destination's nodeId
+     * @param payload       A System::PacketBuffer with the payload. This function takes ownership of the System::PacketBuffer
+     *
+     * @return CHIP_ERROR_NO_MEMORY if no ExchangeContext is available.
+     *         Other CHIP_ERROR codes as returned by the lower layers.
+     */
+    CHIP_ERROR SendEchoRequest(uint64_t nodeId, System::PacketBuffer * payload);
+
+    typedef void (*EchoFunct)(uint64_t nodeId, System::PacketBuffer * payload);
+    EchoFunct OnEchoResponseReceived;
+
+private:
+    ExchangeManager * ExchangeMgr = nullptr;
+    ExchangeContext * ExchangeCtx = nullptr;
+
+    CHIP_ERROR SendEchoRequest(System::PacketBuffer * payload);
+    void OnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType,
+                           System::PacketBuffer * payload);
+};
+
+class DLL_EXPORT EchoServer
+{
+public:
+    EchoServer(void);
+
+    /**
+     *  Initialize the EchoServer object. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    exchangeMgr    A pointer to the ExchangeManager object.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(ExchangeManager * exchangeMgr);
+
+    /**
+     *  Shutdown the EchoServer. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     *  @return #CHIP_NO_ERROR unconditionally.
```

Comment:
why do we return a CHIP_ERROR if we unconditionally succeed?

Reply:
Acked